pkg/asset/installconfig/aws/client.go

```diff
@@ -0,0 +1,144 @@
+package aws
+
+import (
+	"context"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/route53"
+	"github.com/pkg/errors"
+)
+
+//go:generate mockgen -source=./client.go -destination=.mock/awsclient_generated.go -package=mock
+
+// API represents the calls made to the API.
+type API interface {
+	GetPublicHostedZones() (map[string][]string, error)
+	GetPublicHostedZone(zoneName string) (*route53.HostedZone, error)
+	GetPublicHostedZoneByID(zoneID string) (*route53.HostedZone, error)
+}
+
+// Client makes calls to the AWS API.
+type Client struct {
+	meta    *Metadata
+	session *session.Session
+}
+
+// NewClient initializes a client with a session
+func NewClient() (*Client, error) {
+	return &Client{meta: nil, session: nil}, nil
+}
+
+// NewClientWithSession initializes a client with an existing session
+func NewClientWithSession(session *session.Session) (*Client, error) {
+	if session == nil {
+		return nil, errors.Errorf("not a valid session")
+	}
+	return &Client{meta: nil, session: session}, nil
+}
+
+// NewClientWithMetadata initializes a client with AWS metadata
+func NewClientWithMetadata(meta *Metadata) (*Client, error) {
+	if meta == nil {
+		return nil, errors.Errorf("invalid metadata")
+	}
+	return &Client{meta: meta, session: nil}, nil
+}
+
+// getSession will set the session if it's null. This is a hack to allow gomock
+// to run unit tests without the need for a real session.
+func (c *Client) getSession() (*session.Session, error) {
+	if c.session == nil {
+		if c.meta == nil {
+			session, err := GetSession()
+			if err != nil {
+				return nil, err
+			}
+
+			c.session = session
+		} else {
+			session, err := c.meta.Session(context.TODO())
+			if err != nil {
+				return nil, err
+			}
+
+			c.session = session
+		}
+	}
+	return c.session, nil
+}
```

Comment:
```suggestion
// Client makes calls to the AWS API.
type Client struct {
	session *session.Session
}

// NewClient initializes a client with an existing session
func NewClient(session *session.Session) (*Client, error) {
	if session == nil {
		return nil, errors.Errorf("not a valid session")
	}
	return &Client{session: session}, nil
}
```

the other functions should directly use `client.session` instead of calling getSession

Reply:
I've removed the mock framework from AWS.

Reply:
The entire point of /not/ passing in an existing session is to get around the fact that the validation tests don't work when using a session. Any call to GetSession() or meta Session() will cause a validation test to fail. This isn't seen in the Azure and GCP tests, because they don't get run by default. If you run: go test in pkg/asset/installconfig/ it fails because the call to get a public zone ID requires a session, which in turn calls GetSession(). This was a lot of hackery to get around that.